README.md
# simple_calendar_event
Package to create a simple calendar event on all platforms

## Usage
To use this plugin, add ```simple_calendar_event``` as a [dependency in your pubspec.yaml](https://flutter.io/platform-plugins/).

<div align="center">
<img src="https://raw.githubusercontent.com/griajobag/simple_calendar_event/master/screenshot.png"/>
</div>

## Example

```dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';
import 'package:simple_calendar_event/simple_calendar_event.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: "Calendar Event Usage"),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});


  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {


  @override
  Widget build(BuildContext context) {

    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: SimpleCalendarEvent(listEvent: [
          DateTime(2022,09,11),
          DateTime(2022,09,10),
          DateTime(2022,09,15),
          DateTime(2022,09,22),
        ],
          crossAxisSpacing: defaultTargetPlatform == TargetPlatform.android ?1:16,
          mainAxisSpacing:  defaultTargetPlatform == TargetPlatform.android ?1:16,
          onDateClicked: (date){
          // Implement your logic here
            if(kIsWeb){
              Fluttertoast.showToast(
                  msg: date.toString(),
                  toastLength: Toast.LENGTH_SHORT,
                  gravity: ToastGravity.CENTER,
                  timeInSecForIosWeb: 1,
                  backgroundColor: Colors.red,
                  textColor: Colors.white,
                  fontSize: 16.0
              );
            }
          },
        )// This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```
## Other Parameters that may useful

```
paddingBetweenDate // set padding between date
colorOfSelectedDate // Change color of selected date
colorOfEvent // Change color of the circle shape event
isDayEnglish // To set the days in English
isCustomDayText; // To defile that we will use a custom date language
listCustomDay; // If you set isCustomDayText==true, the fullfill this parameter 

```


CHANGELOG.md
## 0.0.1

* Initial publish package.


test/simple_calendar_event_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:simple_calendar_event/simple_calendar_event.dart';

void main() {

}


lib/simple_calendar_event.dart
import 'dart:developer';
import 'package:flutter/material.dart';
import 'package:flutter_staggered_grid_view/flutter_staggered_grid_view.dart';
import 'package:intl/intl.dart';

class SimpleCalendarEvent extends StatefulWidget {
  final List<DateTime> listEvent;
  final Function(DateTime?) onDateClicked;
  final double crossAxisSpacing;
  final double mainAxisSpacing;
  final double paddingBetweenDate;
  final Color colorOfSelectedDate;
  final Color colorOfEvent;
  final bool isDayEnglish;
  final bool isCustomDayText;
  final List<String>? listCustomDay;

  const SimpleCalendarEvent(
      {Key? key,
      required this.listEvent,
      required this.onDateClicked,
      this.crossAxisSpacing = 16,
      this.mainAxisSpacing = 16,
      this.paddingBetweenDate = 10,
      this.colorOfEvent = Colors.blue,
      this.colorOfSelectedDate = Colors.blue,
      this.isDayEnglish = false,
      this.isCustomDayText = false,
      this.listCustomDay})
      : super(key: key);
  @override
  State<SimpleCalendarEvent> createState() => _SimpleCalendarEventState();
}

class _SimpleCalendarEventState extends State<SimpleCalendarEvent> {
  late DateTime lastDayOfCurrentMonth;
  late DateTime lastDayOfPrevMonth;
  late List<String> days;
  late int numberOfLastMonthDate;
  List<DateTime> dateMonthChanges = [];
  List<DateTime> dateDisplayed = [];
  late String monthNameSelected;
  int monthSelected = DateTime.now().month;
  int yearSelected = DateTime.now().year;
  late String dateSelected;

  @override
  void initState() {
    lastDayOfCurrentMonth = DateTime(yearSelected, monthSelected + 1, 0);
    lastDayOfPrevMonth = DateTime(yearSelected, monthSelected - 1, 0);
    numberOfLastMonthDate = DateTime(yearSelected, monthSelected, 0).day;
    _generateDays();
    monthNameSelected = DateFormat("MMM").format(DateTime.now());
    dateSelected = _stringDate(DateTime.now());
    _generatePrevMonthDateForCurrentMonth();
    _generateDisplayedDate();
    _generateNextMonthDateForCurrentMonth();

    super.initState();
  }

  String _stringDate(DateTime date) {
    return DateFormat("dd/MM/yyyy").format(date);
  }

  void _generateDays(){
    setState((){
      if(widget.isCustomDayText && widget.listCustomDay!=null){
        days = widget.listCustomDay!;
      }else{
        if(widget.isDayEnglish){
          days = ["SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT"];
        }else{
          days = ["MIN", "SEN", "SEL", "RAB", "KAM", "JUM", "SAB"];
        }
      }
    });
  }

  void _clearAllList() {
    setState(() {
      dateMonthChanges.clear();
      dateDisplayed.clear();
    });
  }

  void _generateDisplayedDate() {
    for (var i = 1; i <= lastDayOfCurrentMonth.day; i++) {
      dateDisplayed.add(
          DateTime(lastDayOfCurrentMonth.year, lastDayOfCurrentMonth.month, i));
    }
  }

  bool _isContainEvents(DateTime date) {
    bool isEvent = widget.listEvent
        .where((element) => _stringDate(element) == _stringDate(date))
        .isNotEmpty;
    return isEvent;
  }

  void _generateNextMonthDateForCurrentMonth() {
    List<DateTime> listDayNextMonthForCurrentMonth = [];
    for (int i = 0;
        i <
            7 -
                (_totalDisableDatePreviousMonth(
                        DateFormat("E").format(lastDayOfCurrentMonth)) +
                    1);
        i++) {
      listDayNextMonthForCurrentMonth
          .add(DateTime(yearSelected, monthSelected + 1, i + 1));
    }
    dateDisplayed.addAll(listDayNextMonthForCurrentMonth);
  }

  void _generatePrevMonthDateForCurrentMonth() {
    final dateName =
        DateFormat("E").format(DateTime(yearSelected, monthSelected, 1));
    List<DateTime> listDayLastMonthForCurrentMonth = [];
    for (int i = numberOfLastMonthDate;
        i > (numberOfLastMonthDate - _totalDisableDatePreviousMonth(dateName));
        i--) {
      final lastDayKemarin = DateTime(yearSelected, monthSelected - 1, i);
      listDayLastMonthForCurrentMonth.add(
        lastDayKemarin,
      );
    }

    dateDisplayed.addAll(listDayLastMonthForCurrentMonth.reversed);
  }

  int _totalDisableDatePreviousMonth(String day) {
    switch (day.toLowerCase()) {
      case "sun":
        return 0;
      case "mon":
        return 1;
      case "tue":
        return 2;
      case "wed":
        return 3;
      case "thu":
        return 4;
      case "fri":
        return 5;
      case "sat":
        return 6;
      default:
        return 0;
    }
  }

  void _goToPrevMonth() {
    setState(() {
      setState(() {
        _clearAllList();
        if (monthSelected == 1) {
          monthSelected = 12;
          yearSelected = yearSelected - 1;
        } else {
          monthSelected = monthSelected - 1;
        }

        lastDayOfCurrentMonth = DateTime(yearSelected, monthSelected + 1, 0);

        lastDayOfPrevMonth = DateTime(yearSelected, monthSelected - 1, 0);
        numberOfLastMonthDate = DateTime(yearSelected, monthSelected, 0).day;

        monthNameSelected =
            DateFormat("MMM").format(DateTime(yearSelected, monthSelected));
        _generatePrevMonthDateForCurrentMonth();
        _generateDisplayedDate();
        _generateNextMonthDateForCurrentMonth();
      });
    });
  }

  void _goToNextMonth() {
    setState(() {
      _clearAllList();
      if (monthSelected == 12) {
        monthSelected = 1;
        yearSelected = yearSelected + 1;
      } else {
        monthSelected = monthSelected + 1;
      }

      lastDayOfCurrentMonth = DateTime(yearSelected, monthSelected + 1, 0);

      lastDayOfPrevMonth = DateTime(yearSelected, monthSelected - 1, 0);
      numberOfLastMonthDate = DateTime(yearSelected, monthSelected, 0).day;

      monthNameSelected =
          DateFormat("MMM").format(DateTime(yearSelected, monthSelected));
      _generatePrevMonthDateForCurrentMonth();
      _generateDisplayedDate();
      _generateNextMonthDateForCurrentMonth();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
        padding: const EdgeInsets.all(10),
        child: Column(
          children: [
            Container(
              margin: const EdgeInsets.only(bottom: 10),
              padding: const EdgeInsets.all(10),
              width: MediaQuery.of(context).size.width,
              child: Row(
                children: [
                  InkWell(
                    onTap: () {
                      _goToPrevMonth();
                    },
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Row(
                        children: [
                          const Icon(
                            Icons.arrow_back_ios_rounded,
                            size: 15,
                          ),
                          Text("Prev".toUpperCase()),
                        ],
                      ),
                    ),
                  ),
                  Expanded(
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        "${monthNameSelected.toUpperCase()}, ${yearSelected.toString().toUpperCase()}",
                        textAlign: TextAlign.center,
                      ),
                    ),
                  ),
                  InkWell(
                    onTap: () {
                      _goToNextMonth();
                    },
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Row(
                        children: [
                          Text("Next".toUpperCase()),
                          const Icon(
                            Icons.arrow_forward_ios_rounded,
                            size: 15,
                          ),
                        ],
                      ),
                    ),
                  ),
                ],
              ),
            ),
            (widget.isCustomDayText && widget.listCustomDay==null)?Container(
              child: const Text("Your list of day is empty"),
            ): StaggeredGrid.count(
              crossAxisCount: 7,
              children: List.generate(
                  days.length,
                  (index) => Center(
                          child: Text(
                        days[index],
                        style: const TextStyle(fontWeight: FontWeight.bold),
                      ))), // add some space
            ),
            const Padding(padding: EdgeInsets.only(top: 10, bottom: 10)),
            AlignedGridView.count(
              shrinkWrap: true,
              crossAxisCount: 7,
              crossAxisSpacing: widget.crossAxisSpacing,
              mainAxisSpacing: widget.mainAxisSpacing,
              itemCount: dateDisplayed.length,
              itemBuilder: (context, index) => InkWell(
                onTap: () => {
                  widget.onDateClicked(dateDisplayed[index]),
                  setState(() {
                    dateSelected = _stringDate(dateDisplayed[index]);
                    if (dateDisplayed[index].year == yearSelected) {
                      if (monthSelected > dateDisplayed[index].month) {
                        _goToPrevMonth();
                      } else if (monthSelected < dateDisplayed[index].month) {
                        _goToNextMonth();
                      }
                    } else {
                      yearSelected = dateDisplayed[index].year;
                      if (monthSelected == 12) {
                        _goToNextMonth();
                      } else if (monthSelected == 1) {
                        _goToPrevMonth();
                      } else {
                        if (monthSelected > dateDisplayed[index].month) {
                          _goToPrevMonth();
                        } else if (monthSelected < dateDisplayed[index].month) {
                          _goToNextMonth();
                        }
                      }
                    }
                  }),
                },
                child: Container(
                  alignment: Alignment.center,
                  padding: EdgeInsets.all(widget.paddingBetweenDate),
                  child: Column(
                    children: [
                      Text(
                        "${dateDisplayed[index].day}",
                        style: TextStyle(
                            fontWeight: _stringDate(dateDisplayed[index]) ==
                                    _stringDate(DateTime.now())
                                ? FontWeight.bold
                                : FontWeight.normal,
                            color: dateSelected ==
                                    _stringDate(dateDisplayed[index])
                                ? widget.colorOfSelectedDate
                                : dateDisplayed[index].month == monthSelected
                                    ? Colors.black
                                    : Colors.grey),
                      ),
                      _isContainEvents(dateDisplayed[index])
                          ? Icon(
                              Icons.circle,
                              size: 5,
                              color: widget.colorOfEvent,
                            )
                          : Container(
                              height: 5,
                            )
                    ],
                  ),
                ),
              ),
            )
          ],
        ));
  }
}


example/lib/main.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';
import 'package:simple_calendar_event/simple_calendar_event.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: "Calendar Event Usage"),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: SimpleCalendarEvent(
          listEvent: [
            DateTime(2022, 09, 11),
            DateTime(2022, 09, 10),
            DateTime(2022, 09, 15),
            DateTime(2022, 09, 22),
          ],
          crossAxisSpacing:
              defaultTargetPlatform == TargetPlatform.android ? 1 : 16,
          mainAxisSpacing:
              defaultTargetPlatform == TargetPlatform.android ? 1 : 16,
          onDateClicked: (date) {
            if (kIsWeb) {
              Fluttertoast.showToast(
                  msg: date.toString(),
                  toastLength: Toast.LENGTH_SHORT,
                  gravity: ToastGravity.CENTER,
                  timeInSecForIosWeb: 1,
                  backgroundColor: Colors.red,
                  textColor: Colors.white,
                  fontSize: 16.0);
            }
          },
          colorOfSelectedDate: Colors.red,
        ) // This trailing comma makes auto-formatting nicer for build methods.
        );
  }
}
